### PR TITLE
Optimisation

### PR DESCRIPTION
**Pull request by paulocheque**
Just for reference. It is not complete yet

### CODE REVIEW

**Comment by durden** [comment]
Oops did I break scroll sync when I made it the default?  I was trying to keep the UI updated so thought clicking the button programmatically was the best approach.  What did [6a7bc71](https://github.com/pluralsight/guides-cms/pull/47/commits/6a7bc71d07383aad91cce86c260df0ea23527c89) fix?  Just curious so I can avoid breaking it in the future ;)

**Comment by paulocheque** [comment]
I am trying two different approaches: using riot.js and virtual-dom. 

riot.js: I was able to work with Riot.js but using a not smart approach, so the code was even slower. I am unable to work with native tags yet.

virtual-dom: I am still trying with it. I need to create standalone libs for that, so I am experimenting some alternatives. For now I got a 400kb lib without success, so I am trying another libs.

**Comment by paulocheque** [comment]
In this last commit (using virtual-dom), I removed some of the features to isolate the performance problem.

ps: I am using virtual-dom and html-to-vdom libs

ps2: I tried riot.js without success: or bad performance or not working properly (riot.js appears to not use virtual-dom although it claims to)

ps3: I have also tried: virtual-html, virtualize, dom2hscript instead of html-to-vdom. Now I will try vdom-parser like suggested in here: https://github.com/Matt-Esch/virtual-dom/issues/349

**Comment by paulocheque** [comment]
@durden Sorry I forgot to answer your message. 

No, you did not break anything. I broke it during the optimisation.

When we remove the "maxLines", we fix the performance but the scroll sync stop working.

**Comment by paulocheque** [comment]
vdom-parser is significant faster than html-to-vdom, but it is not enough yet

**Comment by paulocheque** [comment]
New observation:

the update/preview is very fast if the guy write the article from top to bottom. I mean, the "edited line" is always the last one, which should be the default behavior. (https://github.com/Matt-Esch/virtual-dom/issues/153)

But if the guy edit a line in the middle of the article (not in the last line) so we have slowness.

I understand the problem and the virtual dom has mechanisms to fix that. Now I am trying to figure it out the solution. (https://github.com/chjj/marked/issues/743)

File: pskb_website/static/js/editor_utils.js [hunk 1]
```diff
@@ -248,11 +247,25 @@ function initialize_editor(local_filename, content, name, real_name, img_upload_
         }
     }, 1000));
 
+    editor.getSession().on('changeScrollTop', function(scrollTop) {
```
Comment (durden) [inline pskb_website/static/js/editor_utils.js]: Very nitpick issue here, but seems like the standard in this file is to use semicolons so makes sense to add them here too.  Would we get anything from marking this file with `use strict;`?

**Comment by durden** [comment]
This looks much better.  I tested it with [this large guide](http://guides-dev.herokuapp.com/other/big-js-guide-to-test-with?status=in-review). I think this was the guide the issue was first reported with.

I saw rendering times mostly hover around 300ms when changing data all over the guide.

The scrolling is still a big laggy but I think it's acceptable for a large guide.

I haven't looked too close into the implementation, but could we improve the rendering by increasing the time between rendering?  For example, if the user is typing very fast could we increase the time between updates and only refresh the preview after the typing has stopped for x amount of ms?

Just a thought that we might be updating on every keypress and instead we could coalesce a bunch of small changes into 1 update.

**Comment by paulocheque** [comment]
Cool. After finishing the optimisation, I can make an extra commit to fix all code style. Adding the strict mode. Right now I am making too many tests that I am forgetting to add ;s

The number of DOM nodes updated is linearly related with the performance. For now, we can see that the number of nodes updates for each edition is basically the number of lines in the article below the edition (below the current cursor position).

Example with n lines:

```
line 1: aaa (if we add a line below aaa, it will result in **n** updates: new node, ..., bbb, ccc, ddd => slow)
line 2: bbb (if we add a line below bbb, it will result in **n-1** updates: new node, ccc, ddd => slow)
...
line n-1: ccc (if we add a line below ccc, it will result in 2 updates: new node, ddd => fast)
line n: ddd (if we add a line below ddd, it will result in 1 update: new node => fast)
```

Yes, we can increase the time to update the preview. I will try to let this time "dynamic" to be smarter according to the cursor position.

**Comment by paulocheque** [comment]
screen shot 2016-04-24 at 4 42 29 am

I tried `virtual-markdown` to avoid one unnecessary step (convert markdown to an html string), but the results were not satisfactory: 93 + 37 < 277

**Comment by paulocheque** [comment]
I fixed the highlight performance, fixed some html code issues. Added target=_blank to links, removed logs etc. I think we can merge it after some final tests. I will rebase the branch with the current master right now.

**Comment by paulocheque** [comment]
Well, there were lots and lots of conflicts between the branches. I rebases just the initial commits without conflicts in the `optimisation` branch (the branch of this PR). 

To avoid lost anything, I created another branch called `optimisation2`. I rebased all commits with the `master` branch: https://github.com/paulocheque/guides-cms/commits/optimisation2

Then, I deployed the `optimisation2` branch in my dev env for tests. Please, take a look.

**Comment by durden** [comment]
I tested your `optimisation2` branch on your testing site.  It looks good!  The scrolling is super fast now!

I did find a few UI issues that are probably related to the rebase:
- Scroll sync is automatically enabled on launch but menu option doesn't have the checkbox on it by default
  - Essentially the checkbox is out of sync with the enabling of the action itself
- There's still something wrong with `z-index` and the editor page.
  - Try clicking 'save' on a guide without entering a title to see an example.
  - The modal popup shows up but you cannot exit it in anyway.

**Comment by durden** [comment]
I think the performance on this branch is ready to push to the production server.  Can you fix the small issues, rebase against master, and update this PR?

I can then pull the clean PR locally, test it, merge, and we can test once more on the staging site with all the changes.

**Comment by paulocheque** [comment]
If the underscore team accepts this PR: https://github.com/jashkenas/underscore/pull/2518, we can update the lib and remove our debounce function.

**Comment by paulocheque** [comment]
Fixed the checkbox issue. Also, I include the PR/commit https://github.com/pluralsight/guides-cms/pull/55 in the last rebase.

**Comment by paulocheque** [comment]
It is all in the `optimisation` branch right now. I am not using the `optimisation2` anymore since most of the things appears to be ok.

File: pskb_website/static/js/editor_utils.js [hunk 2]
```diff
@@ -407,7 +407,7 @@ function openFullscreen() {
 function resizeEditor() {
     var lineHeight = editor.renderer.lineHeight;
     var maxLines = document.getElementById('editor-wrapper').offsetHeight / lineHeight;
-    editor.setOption('maxLines', Math.floor(maxLines) - 1);
```
Comment (durden) [inline pskb_website/static/js/editor_utils.js]: Wasn't the maxLines option causing some of the performance issues?  Why was there a `- 1` to begin with? Just curious.

**Comment by durden** [comment]
I merged these changes and pushed the master branch to [this testing site](http://pskb-stage.herokuapp.com).  Everything looks good so far. @paulocheque could you run through your usual tests on the test URL just to make sure everything looks OK to you?  I'll push to production once you also verify it looks ok.  Thanks!

File: pskb_website/static/js/editor_utils.js (hunk 2 above)
Comment (paulocheque) [inline pskb_website/static/js/editor_utils.js]: For the scroll sync to work with the past implementation, we need to set a high number in the maxLines. We had set something like 50000 or something. https://github.com/ajaxorg/ace/issues/2947

**Comment by paulocheque** [comment]
cool! yes, I will check.

**Comment by paulocheque** [comment]
Everything is fine! cool!

**Comment by durden** [comment]
Thanks, it's all up on production now!

**Comment by paulocheque** [comment]
PS about some js libs:

Since we are not using any js framework to import another js files, like Node, CommonJS, RequireJS, ECMA2015 etc, we need to customize the libraries to use with raw javascript.

Example:

`index-vdom-parser.js`:

```
var vdomParser = require('vdom-parser')
module.exports = vdomParser;
```

`browserify index-vdom-parser.js --standalone vdomParser > dist/vdom-parser-standalone.js`

**Pull request merged**